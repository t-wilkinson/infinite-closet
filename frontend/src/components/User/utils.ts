import { StrapiUser } from '@/types/models'

export const displayUsername = (username: string) => {
  return username ? username.split('-').slice(0, -1).join('-') : ''
}
