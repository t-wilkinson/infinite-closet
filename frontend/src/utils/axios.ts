import Axios from 'axios'

export const axios = Axios.create({
  baseURL: process.env.NEXT_PUBLIC_BACKEND,
  headers: {
    'Content-Type': 'application/json',
  },
  withCredentials: true,
})

axios.interceptors.response.use(
  (response): any => {
    return response.data
  },
  (error) => {
    // const message = error.response?.data?.message || error.message;
    // useNotificationStore.getState().addNotification({
    //   type: 'error',
    //   title: 'Error',
    //   message,
    // });
    return Promise.reject(error)
  }
)

export function getURL(url: string) {
  if (url == null) {
    return null
  }

  if (url.startsWith('http') || url.startsWith('//')) {
    return url
  }

  return `${process.env.NEXT_PUBLIC_BACKEND}${url}`
}

export default axios
