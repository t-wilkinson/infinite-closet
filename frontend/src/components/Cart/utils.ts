import { createAsyncThunk } from '@reduxjs/toolkit'

import axios from '@/utils/axios'
import { RootState } from '@/utils/store'
import { StrapiOrder } from '@/utils/models'
import {Summary} from '@/types'

import * as helpers from './helpers'
import { Orders, Cart } from './types'

export const getUser = (getState: () => any) =>
  (getState() as RootState).user.data

const toKey = (order: StrapiOrder) => `${order.size}_${order.product.id}`

export default {
  get: createAsyncThunk<Orders, void>('cart/get', async (_, { getState }) => {
    const user = getUser(getState)
    return await helpers.getOrders(user)
  }),
  set: createAsyncThunk<Orders, Orders>(
    'cart/set',
    async (orders, { getState }) => {
      const user = getUser(getState)
      const uniqueOrders: Orders = Object.values(
        orders.reduce((acc, order) => {
          acc[toKey(order)] = order
          return acc
        }, {})
      )
      helpers.setOrders(user, uniqueOrders)
      return orders
    }
  ),
  summary: createAsyncThunk<Summary, void>(
    'cart/summary',
    async (_, { getState }) => {
      const user = getUser(getState)
      const orders = await helpers.getOrders(user)
      let data: Summary
      if (user) {
        data = await axios.post<Summary>(
          `/orders/cart/summary`,
          { orders },
        )
      } else {
        data = await axios.post(`/orders/cart/summary`, { orders }, {withCredentials: false})
      }
      return data
    }
  ),
  count: createAsyncThunk<number, void>(
    'cart/count',
    async (_, { getState }) => {
      const user = getUser(getState)
      if (user) {
        return await axios.get<number>('/orders/cart/count')
      } else {
        const orders = helpers.getGuestOrders()
        return orders.length
      }
    }
  ),
  view: createAsyncThunk<Cart, void>('cart/view', async (_, { getState }) => {
    const user = getUser(getState)
    return helpers.viewOrders(user)
  }),
  history: createAsyncThunk<Cart, void>(
    'cart/history',
    async (_, { getState }) => {
      const user = getUser(getState)
      return helpers.orderHistory(user)
    }
  ),
  update: createAsyncThunk<void, Partial<StrapiOrder>>(
    'cart/update',
    async (order, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.put<void>(`/orders/${order.id}`, order)
      } else {
        await axios.put<void>(`/orders/${order.id}`, order, {withCredentials: false})
        const orders = helpers.getGuestOrders()
        const index = orders.findIndex((v: StrapiOrder) => v.id === order.id)
        orders[index] = { ...orders[index], ...order }
        helpers.setGuestOrders(orders)
      }
    }
  ),
  add: createAsyncThunk<Cart, Partial<StrapiOrder>>(
    'cart/add',
    async (order, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.post<void>('/orders', order)
      } else {
        order = await axios.post<StrapiOrder>('/orders', order, {withCredentials: false})
        const orders = [...helpers.getGuestOrders(), order as StrapiOrder]
        helpers.setGuestOrders(orders)
      }
      return await helpers.viewOrders(user)
    }
  ),
  insert: createAsyncThunk<Cart, Partial<StrapiOrder>[] | Partial<StrapiOrder>>(
    'cart/insert',
    async (inserts, { getState }) => {
      const user = getUser(getState)
      if (!user) {
        return
      }

      if (!Array.isArray(inserts)) {
        inserts = [inserts]
      }

      const orders = await helpers.getUserOrders(user)
      const orderIds = orders
        .concat(inserts as Orders)
        .map((order: StrapiOrder) => order.id)
      helpers.setUserOrders(user, orderIds)
      return await helpers.viewOrders(user)
    }
  ),
  remove: createAsyncThunk<Orders, string>(
    'cart/remove',
    async (id, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.put<void>(`/orders/${id}`, { id, status: 'dropped' })
      } else {
        const orders = helpers.getGuestOrders()
        const filtered = orders.filter((order: StrapiOrder) => order.id !== id)
        helpers.setGuestOrders(filtered)
      }
      return helpers.getOrders(user)
    }
  ),
}
