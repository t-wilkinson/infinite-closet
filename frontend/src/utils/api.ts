import axios from 'axios'

export function getURL(url: string) {
  if (url == null) {
    return null
  }

  if (url.startsWith('http') || url.startsWith('//')) {
    return url
  }

  return `${process.env.NEXT_PUBLIC_BACKEND}${url}`
}

// Helper to make GET requests to Strapi
export async function fetchAPI(path: string) {
  const { data } = await axios.get(path, {
    withCredentials: true,
    headers: {
      'Content-Type': 'application/json',
    },
  })
  return data
}
