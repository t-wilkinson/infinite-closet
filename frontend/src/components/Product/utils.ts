import { getURL } from '@/utils/axios'
import { StrapiProduct } from '@/types/models'

export const productImageProps = (product: StrapiProduct) => {
  const mainImage = product.images[0]
  return {
    src: getURL(mainImage.formats.thumbnail?.url || mainImage.url),
    alt:
      mainImage.alternativeText ||
      product.designer ? `${product.name} by ${product.designer.name}` : product.name,
  }
}

export const productRentalPrice = (product: StrapiProduct, rentalLength: 'short' | 'long') => {
  switch (rentalLength) {
    case 'short':
      return product.shortRentalPrice
    case 'long':
      return product.longRentalPrice
    default:
      return undefined
  }
}
