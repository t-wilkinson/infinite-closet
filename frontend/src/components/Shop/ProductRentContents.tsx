import React from 'react'
import axios from 'axios'
import { useRouter } from 'next/router'

import useAnalytics from '@/utils/useAnalytics'
import { useDispatch, useSelector } from '@/utils/store'
import { Icon } from '@/components'
import { Submit } from '@/Form'
import { toTitleCase } from '@/utils/helpers'
import { unNormalizeSize } from '@/Products/helpers'

import { shopActions } from './slice'
import { OneTime } from './types'
import DatePicker from './DatePicker'

export const ProductRentContents = ({ data, product, state }) => {
  const router = useRouter()
  const Contents = productRentContents[state.rentType]
  const user = useSelector((state) => state.user.data)
  const dispatch = useDispatch()

  return (
    <div className="">
      <Contents
        data={data}
        router={router}
        user={user}
        product={product}
        state={state}
        dispatch={dispatch}
      />
    </div>
  )
}
export default ProductRentContents

const rentalLengths: { [key in OneTime]: number } = {
  Short: 4,
  Long: 8,
}

interface Size {
  id: number
  size: string
  quantity: number
}

export const productRentContents = {
  OneTime: ({ data, user, dispatch, product, state, router }) => {
    const [status, setStatus] = React.useState<null | string>(null)
    const [chartOpen, setChartOpen] = React.useState(false)
    const analytics = useAnalytics()

    const addToCart = () => {
      setStatus('adding')
      axios
        .post(
          '/orders',
          {
            status: 'cart',
            size: unNormalizeSize(product.sizes[state.size].size),
            date: state.selectedDate.toJSON(),
            rentalLength: state.oneTime,
            product: product.id,
          },
          { withCredentials: true },
        )
        .then(() => {
          router.push('/user/checkout')
          analytics?.logEvent('add_to_cart', {
            user: user.email,
          })
        })
        .catch((err) => {
          console.error(err)
          setStatus('error')
        })
    }

    return (
      <>
        <DatePicker
          state={state}
          rentalLength={rentalLengths[state.oneTime]}
          dispatch={dispatch}
        />

        <SelectorItem label="Size" className="my-2 z-10 w-full">
          {chartOpen && (
            <SizeChart
              measurements={data.sizeChart.measurements}
              product={product}
              chart={data.sizeChart.chart}
              sizeEnum={data.sizeChart.sizeEnum}
              close={() => setChartOpen(false)}
            />
          )}
          <div className="relative flex-row justify-start space-x-4 items-center w-full">
            {/* select elements are too difficult to style
                divs don't act like buttons
                buttons can't use aria-role
            */}
            <SizeSelector dispatch={dispatch} product={product} state={state} />
            <button onClick={() => setChartOpen((state) => !state)}>
              <span className="underline">Size Chart</span>
            </button>
          </div>
        </SelectorItem>

        <SelectorItem label="Rental time" className="my-2">
          <div className="flex-row justify-between w-full flex-wrap">
            <div className="mr-6">
              <OneTimeRadioButton
                selected={state.oneTime === 'Short'}
                oneTime="Short"
                dispatch={dispatch}
              />
              <OneTimeRadioButton
                selected={state.oneTime === 'Long'}
                oneTime="Long"
                dispatch={dispatch}
              />
            </div>
            <button
              className="flex flex-grow border border-gray py-2 px-2 rounded-sm rounded-sm flex-row flex-grow justify-between items-center"
              onClick={() => dispatch(shopActions.showDate())}
            >
              <span>
                {state.selectedDate &&
                  state.selectedDate.format('ddd M/D') +
                    ' - ' +
                    state.selectedDate
                      .add(rentalLengths[state.oneTime], 'day')
                      .format('ddd M/D')}
              </span>
              <Icon className="text-gray" name="date" size={24} />
            </button>
          </div>
        </SelectorItem>

        <Submit
          onSubmit={addToCart}
          className="my-2 self-center rounded-sm w-full"
          disabled={
            process.env.NODE_ENV === 'production' ||
            !state.selectedDate ||
            !user ||
            state.size === undefined ||
            status === 'adding'
          }
        >
          {process.env.NODE_ENV === 'production'
            ? 'Coming Soon'
            : !user
            ? 'Please sign in'
            : status === 'adding'
            ? 'Adding...'
            : status === 'error'
            ? 'Unable to add to cart'
            : 'Add to Cart'}
        </Submit>
      </>
    )
  },

  Membership: () => (
    <div className="justify-center items-center flex-grow h-64">
      <span className="font-subheader text-center text-3xl">Coming Soon</span>
    </div>
  ),

  Purchase: () => (
    <div className="justify-center items-center flex-grow h-64">
      <span className="font-subheader text-center text-3xl">Coming Soon</span>
    </div>
  ),
}

const SizeSelector = ({ product, state, dispatch }) => {
  const [sizeState, setSizeState] = React.useState(false)

  return (
    <div>
      <button
        tabIndex={0}
        aria-label="Dropdown product sizes"
        className="flex p-2 border border-gray relative cursor-pointer w-32 justify-between flex-row"
        onClick={() => setSizeState((state) => !state)}
      >
        {(state.size !== undefined &&
          product.sizes[state.size] !== undefined &&
          product.sizes[state.size].size) ||
          'Select Size'}
        <Icon name="down" size={16} className="mt-1" />
      </button>
      <div
        className={`
        w-32 absolute bottom-0 bg-white divide-y transform translate-y-full border border-gray z-10
        ${sizeState ? '' : 'hidden'}
        `}
      >
        {product.sizes.map((size: Size, index: number) => (
          <button
            key={size.id}
            tabIndex={0}
            aria-label="Dropdown sizes"
            onClick={() => {
              dispatch(shopActions.changeSize(index))
              setSizeState(false)
            }}
            className="flex justify-center cursor-pointer bg-white"
          >
            {size.size}
          </button>
        ))}
      </div>
    </div>
  )
}

const SizeChart = ({ product, sizeEnum, chart, close, measurements }) => {
  return (
    <div
      className="absolute bg-white border-gray border p-4 z-10 pt-8 space-y-4 overflow-y-scroll"
      style={{ maxHeight: 600 }}
    >
      <button onClick={close} className="absolute top-0 right-0">
        <div className="p-4">
          <Icon name="close" size={16} />
        </div>
      </button>

      <table className="table-fixed border border-gray-light">
        <thead className="border border-gray-light">
          <tr className="border-b border-gray-light">
            <th colSpan={chart.length + 1}>Womens Clothing</th>
          </tr>
          <tr className="border-b border-gray-light">
            <th scope="col" />
            {chart.map((item) => (
              <th key={item.name} scope="col" className="w-12">
                {item.name}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="p-2">
          {sizeEnum.map((size) => (
            <tr key={size} className="border-t border-gray-light">
              <th scope="row" className="p-1">
                {size}
              </th>
              {chart.map((item) => (
                <td key={item.name} className="text-center">
                  {item[size] || '-'}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>

      <table className="table-fixed border border-gray-light">
        <thead className="border border-gray-light">
          <tr className="border-b border-gray-light">
            <th colSpan={chart.length + 1}>{product.name} Measurements</th>
          </tr>
          <tr className="border-b border-gray-light">
            <th scope="col" />
            {measurements.map((measurement) => (
              <th key={measurement} className="text-center">
                {toTitleCase(measurement)}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="p-2">
          {product.sizes.map((size) => (
            <tr key={size.id} className="border-t border-gray-light">
              <th scope="row" className="p-1">
                {size.sizeRange ? `${size.size}/${size.sizeRange}` : size.size}
              </th>
              {measurements.map((measurement) => (
                <td key={measurement} className="text-center">
                  {size[measurement] || '-'}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  )
}

const SelectorItem = ({ label, children, ...props }) => (
  <div {...props}>
    <span className="font-bold my-2">{label}</span>
    {children}
  </div>
)

const OneTimeRadioButton = ({ selected, oneTime, dispatch }) => (
  <button
    className="flex-row flex items-center"
    onClick={() => dispatch(shopActions.changeOneTime(oneTime))}
  >
    <div className="w-4 h-4 rounded-full border border-gray items-center justify-center mr-2">
      <div
        className={`w-3 h-3 rounded-full
        ${selected ? 'bg-sec-light' : ''}
        `}
      />
    </div>
    {/* TODO: dynamic from server */}
    <span>{{ Short: 4, Long: 8 }[oneTime]}-day rental</span>
  </button>
)
