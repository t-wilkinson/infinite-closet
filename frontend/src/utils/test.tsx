import React from 'react'
import { configureStore } from '@reduxjs/toolkit'
import { storeOptions } from '@/utils/store'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'

function render(
  ui: React.ReactElement<any>,
  {
    initialState = {},
    store = configureStore({ ...storeOptions, preloadedState: initialState }),
    ...renderOptions
  } = {},
) {
  store.dispatch = jest.fn()

  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  let component = rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
  component.store = store
  return component
}

export * from '@testing-library/react'
export { render }
