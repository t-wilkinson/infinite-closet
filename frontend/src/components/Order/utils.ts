import { toast } from 'react-toastify'

import axios from '@/utils/axios'
import * as sizing from '@/utils/sizing'
import { currency } from '@/utils/config'
import { productRentalPrice } from '@/Product/utils'

import { OrderUtils } from './slice'

const prepareOrder = ({ product, fields, user }) => {
  const size = sizing.get(product.sizes, fields.get('size').value)
  const selectedDate = fields.get('selectedDate').value
  const rentalLength = fields.get('rentalLength').value

  const order: any = {
    user: user ? user.id : null,
    size: sizing.unnormalize(size.size),
    product: product.id,
    expectedStart: selectedDate ? selectedDate.toJSON() : undefined,
    rentalLength: rentalLength ? rentalLength : undefined,
  }
  return order
}

const productToItem = (product) => {
  return {
    item_id: product.id,
    item_name: product.name,
    item_brand: product.designer.name,
    // item_category: order.,
    price: product.retailPrice,
    quantity: 1,
  }
}

export const addToCart = async ({
  product,
  fields,
  user,
  dispatch,
  analytics,
}) => {
  const order = { ...prepareOrder({ product, fields, user }), status: 'cart' }

  return dispatch(OrderUtils.add(order))
    .then(() => {
      toast.success(`Successfully added to cart.`, {
        autoClose: 1500,
        hideProgressBar: true,
      }),
        analytics.logEvent('add_to_cart', {
          currency,
          value: productRentalPrice(product, order.rentalLength),
          user: user ? user.email : 'guest',
          items: [
            productToItem(product)
          ],
        })
    })
    .catch(() => {
      toast.error(`Unable to add to cart`, {})
    })
}

export const addToFavorites = async ({
  product,
  fields,
  user,
  dispatch,
  analytics,
}) => {
  try {
    const order = { ...prepareOrder({ product, fields, user }), status: 'list' }
    await dispatch(OrderUtils.add(order))
    await dispatch(OrderUtils.favorites())
    toast.success(`Successfully added to favorites.`, {
      autoClose: 1500,
      hideProgressBar: true,
    })
    analytics.logEvent('add_to_wishlist', {
      currency,
      value: productRentalPrice(product, order.rentalLength),
      items: [
         productToItem(product)
      ],
    })
  } catch (e) {
    console.log(e)
    toast.error(
      `Ran into an issue adding to favorites. We'll have this fixed soon!`,
      {}
    )
  }
}

export const removeOrderItem = async ({ dispatch, analytics, order }) => {
  const { product } = order
  await dispatch(OrderUtils.remove(order.id))
  await dispatch(OrderUtils.view())
  analytics.logEvent('remove_from_cart', {
    currency,
    value: productRentalPrice(product, order.rentalLength),
    items: [
      productToItem(product)
    ],
  })
}

export const toggleInsurance = ({ dispatch, order }) => {
  dispatch(
    OrderUtils.update({ id: order.id, insurance: !order.insurance })
  ).then(() => dispatch(OrderUtils.view()))
}

export const moveToFavorites = ({ order, dispatch }) => {
  dispatch(OrderUtils.update({ id: order.id, status: 'list' })).then(() => {
    dispatch(OrderUtils.view())
    dispatch(OrderUtils.favorites())
  })
}

export const getValidOrderDates = ({ product, dates, size, rentalLength }) =>
  axios
    .post(
      '/orders/dates/valid',
      {
        dates,
        product: product.id,
        size: (sizing.get(product.sizes, size) || product.sizes[0]).size,
        rentalLength,
      },
      { withCredentials: false }
    )
    .then((data) => data.valid)
    .catch((err) => {
      toast.error(`We ran into an issue verifying the available dates.`)
    })

export const availableOrderDates = async ({
  date,
  product,
  size,
  rentalLength,
}) => {
  const curDay = date.hour(12).date(1)
  const dates = [
    curDay.date(0).subtract(1, 'month'),
    curDay.date(0),
    curDay.add(1, 'month').date(0),
    curDay.add(2, 'month').date(0),
    curDay.add(3, 'month').date(0),
  ].reduce((acc, day) => {
    Array(day.date())
      .fill(0)
      .map((_, i) => {
        acc.push(day.set('date', i + 1))
      })
    return acc
  }, [])

  // TODO: should not refetch everytime new date overed over
  return await getValidOrderDates({ product, size, rentalLength, dates })
}

const rentalLengths = {
  short: 4 - 1,
  long: 8 - 1,
}

export const getRentalLength = (rentalLength: 'short' | 'long') =>
  rentalLengths[rentalLength]
