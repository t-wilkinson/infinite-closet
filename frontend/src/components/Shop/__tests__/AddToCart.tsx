import React from 'react'
import ProductRentContents from '../ProductRentContents'
import { render, fireEvent, RenderResult, within } from '@/utils/test'

const mockProduct = {
  id: 'string',
  name: 'string',
  slug: 'string',
  designer: { name: 'Designer' },
  shortRentalPrice: 10.0,
  longRentalPrice: 10.0,
  retailPrice: 10.0,
  purchasePrice: 10.0,
  images: [],
  sizes: [{ id: '1', quantity: 1, size: 'MD' }],
  details: 'string',
  stylistNotes: 'string',
  sizingNotes: 'string',
  categories: [],
  colors: [],
  fits: [],
  occasions: [],
  styles: [],
  weather: [],
  categories_: 'string',
  colors_: 'string',
  fits_: 'string',
  occasions_: 'string',
  styles_: 'string',
  weather_: 'string',
  created_by: 'string',
  updated_at: 'string',
  published_at: 'string',
}

const mockState = {
  rentType: 'OneTime',
  oneTime: 'Short',
  membership: 'Short',
  dateVisible: false,
}

describe('Shop product', () => {
  let component: RenderResult

  beforeEach(() => {
    component = render(
      <ProductRentContents product={mockProduct} state={mockState} />,
    )
  })

  it('renders correctly', () => {
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('can add size', () => {
    const Size = component.getByText('Size').nextElementSibling as HTMLElement
    const SizeDropDown = within(Size).getByLabelText('Dropdown product sizes')
    fireEvent.click(SizeDropDown, {})
    const sizeMD = within(Size).getByText(/md/i)
    fireEvent.click(sizeMD)
    expect(component.store.dispatch).toHaveBeenCalled()
    // expect(component.asFragment()).toMatchSnapshot()
  })

  // it('is clicked', () => {
  //   renderer.act(() => {
  //     component.root.findByType('button').props.onClick()
  //   })
  //   expect(store.dispatch).toHaveBeenCalled()
  // })
})
