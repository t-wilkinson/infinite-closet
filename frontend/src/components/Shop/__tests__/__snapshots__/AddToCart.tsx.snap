// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop product renders correctly 1`] = `
<DocumentFragment>
  <div
    class="h-96"
  >
    <div
      class="my-2 z-10 w-full"
    >
      <span
        class="font-bold my-2"
      >
        Size
      </span>
      <div
        class="relative items-start w-full"
      >
        <button
          aria-label="Dropdown product sizes"
          class="flex p-2 border border-gray relative cursor-pointer w-32 justify-between flex-row"
        >
          Select Size
          <div
            class="fill-current mt-1"
            style="width: 16px; height: 16px;"
          >
            <svg
              viewBox="0 0 451.847 451.847"
            >
              <path
                d="M225.923,354.706c-8.098,0-16.195-3.092-22.369-9.263L9.27,151.157c-12.359-12.359-12.359-32.397,0-44.751 c12.354-12.354,32.388-12.354,44.748,0l171.905,171.915l171.906-171.909c12.359-12.354,32.391-12.354,44.744,0 c12.365,12.354,12.365,32.392,0,44.751L248.292,345.449C242.115,351.621,234.018,354.706,225.923,354.706z"
              />
            </svg>
          </div>
        </button>
        <div
          class="
              w-32 absolute bottom-0 bg-white divide-y transform translate-y-full border border-gray z-10
              hidden
              "
        >
          <button
            aria-label="Dropdown sizes"
            class="flex justify-center cursor-pointer bg-white"
          >
            MD
          </button>
        </div>
      </div>
    </div>
    <div
      class="my-2"
    >
      <span
        class="font-bold my-2"
      >
        Rental time
      </span>
      <div
        class="flex-row justify-between w-full flex-wrap"
      >
        <div
          class="mr-6"
        >
          <button
            class="flex-row flex items-center"
          >
            <div
              class="w-4 h-4 rounded-full border border-gray items-center justify-center mr-2"
            >
              <div
                class="w-3 h-3 rounded-full
        bg-sec-light
        "
              />
            </div>
            <span>
              4-day rental
            </span>
          </button>
          <button
            class="flex-row flex items-center"
          >
            <div
              class="w-4 h-4 rounded-full border border-gray items-center justify-center mr-2"
            >
              <div
                class="w-3 h-3 rounded-full
        
        "
              />
            </div>
            <span>
              8-day rental
            </span>
          </button>
        </div>
        <button
          class="flex flex-grow border border-gray py-2 px-2 rounded-sm rounded-sm flex-row flex-grow justify-between items-center"
        >
          <span />
          <div
            class="fill-current text-gray"
            style="width: 24px; height: 24px;"
          >
            <svg
              viewBox="0 0 511.977 511.977"
            >
              <path
                d="m14.977 390.988h392c3.516 0 6.914-1.23 9.609-3.472 3.765-3.153 89.652-77.038 94.889-236.528h-421c-5.217 144.774-84.315 212.822-85.137 213.501-4.819 4.072-6.592 10.723-4.424 16.641 2.153 5.903 7.764 9.858 14.063 9.858z"
              />
              <path
                d="m496.976 60.988h-75v-15c0-8.4-6.6-15-15-15s-15 6.6-15 15v15h-76v-15c0-8.4-6.6-15-15-15s-15 6.6-15 15v15h-75v-15c0-8.4-6.6-15-15-15s-15 6.6-15 15v15h-75c-8.4 0-15 6.6-15 15v45h421v-45c0-8.4-6.6-15-15-15z"
              />
              <path
                d="m435.849 410.515c-8.145 6.782-18.369 10.474-28.872 10.474h-316v45c0 8.291 6.709 15 15 15h391c8.291 0 15-6.709 15-15v-168.146c-28.92 70.951-69.276 106.937-76.128 112.672z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <button
      aria-label="Submit form"
      class="p-4 text-white mt-4 rounded-sm
      bg-pri-light
      my-2 self-center rounded-sm w-full
    "
      disabled=""
      type="submit"
    >
      Please sign in
    </button>
  </div>
</DocumentFragment>
`;
